Cython/Compiler/Code.py

```diff
@@ -626,23 +626,29 @@ class UtilityCode(UtilityCodeBase):
     file            filename of the utility code file this utility was loaded
                     from (or None)
     """
+    code_parts = ["proto", "impl", "init", "cleanup", "module_state_decls"]
 
-    def __init__(self, proto=None, impl=None, init=None, cleanup=None, requires=None,
+    def __init__(self, proto=None, impl=None, init=None, cleanup=None,
+                 module_state_decls=None, requires=None,
                  proto_block='utility_code_proto', name=None, file=None):
         # proto_block: Which code block to dump prototype in. See GlobalState.
         self.proto = proto
         self.impl = impl
         self.init = init
         self.cleanup = cleanup
+        self.module_state_decls = module_state_decls
         self.requires = requires
         self._cache = {}
         self.specialize_list = []
         self.proto_block = proto_block
         self.name = name
         self.file = file
 
+    def _parts_tuple(self):
+        return tuple(getattr(self, part, None) for part in self.code_parts)
```

Comment:
Since we're caching utility code quite extensively, the "parts tuple" is almost always used. Let's just calculate it ahead of time in the constructor rather than each time we need the hash value or a comparison.